inference-engine/include/details/ie_cnn_network_iterator.hpp

```diff
@@ -117,7 +120,13 @@ CNNNetworkIterator {
      * @return true if the given iterator is equal to this one, false - otherwise
      */
     bool operator==(const CNNNetworkIterator& that) const {
-        return network == that.network && currentLayer == that.currentLayer;
+        bool retVal;
+        if (currentLayer == nullptr && that.currentLayer == nullptr) {
+            retVal = true;
```

Comment:
What if `network != that.network` and  `network` or/and `that.network` are not equal `nullptr`?

Reply:
The current iterator implementation is not similar to what we would see in std-c++ implementations, it's very customized. Currently typical usage of this iterator is like this:
```
details::CNNNetworkIterator i(&network);
while (i != details::CNNNetworkIterator()) { layers.push_back(*i); i++;  ... }
``` 
The CNNNetworkIterator() = default constructor will init network as nullptr, and it will stay as nullptr; whereas the CNNNetworkIterator(network) constructor will always set network to a value (variable i). The if-logic handles the special case of this.currentLayer being nullptr, which happens when it has exhausted all layers to visit (BTW, that.currentLayer will remain nullptr always in our use case). When network != that.network, the first part of the else-logic will fail, meaning the == operator will return false.